src/optimize/base_optimizer.js

```diff
@@ -102,6 +106,57 @@ export default class BaseOptimizer {
     });
   }
 
+  warmupThreadLoaderPool() {
+    const baseModules = [
+      'babel-loader',
+      BABEL_PRESET_PATH
+    ];
+
+    const devOnlyModules = !IS_KIBANA_DISTRIBUTABLE
+      ? ['ts-loader']
+      : [];
+
+    threadLoader.warmup(
+      // pool options, like passed to loader options
+      // must match loader options to boot the correct pool
+      this.getThreadLoaderPoolConfig(),
+      [
+        // modules to load on the pool
+        ...baseModules,
+        ...devOnlyModules
+      ]
+    );
+  }
+
+  getThreadPoolCpuCount() {
+    const cpus = os.cpus();
+    if (!cpus) {
+      // sometimes this call returns undefined so we fall back to 1: https://github.com/nodejs/node/issues/19022
+      return 1;
+    }
+
+    return Math.max(1, cpus.length - 1);
```

Comment:
We should probably limit the number of processes to something like 5. I tried this out on my 32 core threadripper with a couple different limits and these are the numbers I got:

| max threads | start from clean | start from cache | on change |
| --- | --- | --- | --- |
| 31 | 39.82 | 14.56 | 3.13 |
| 12 | 36.66 | 14.87 | 2.76 |
| 10 | 36.90 | 14.97 | 2.63 |
| 7 | 35.70 | 14.32 | 2.96 |
| 5 | 36.27 | 14.42 | 2.58 |
| 3 | 41.35 | 14.75 | 2.56 |
| 1 | 78.59 | 14.86  | 2.70 |

I only ran these configurations once, so apply some slop to them, but based on these I think spawning 31 workers on my computer is overkill 😄

Reply:
haha @spalger 😅 ! I think we can limit the max as `10` workers, at least for now we shouldn't need more!